### ISSUE TITLE
Why? What? Who? How? (Tutorial) - Please Answer The Key Questions

### ISSUE THREAD

**Issue by nelsonic** [dwyl/learn-insomnia#1]
Hi @th0mas, 👋 
as noted in [`#7`](https://github.com/dwyl/smart-home-security-system/pull/7#discussion_r452921556), [`#8`](https://github.com/dwyl/smart-home-security-system/pull/8#pullrequestreview-446812978) and [`#9`](https://github.com/dwyl/smart-home-security-system/issues/9#issuecomment-656750657),  it would be _really_ good to have a complete beginners guide to [**`insomnia core`**](https://insomnia.rest/products/core/) in _general_ 
and then _specifically_ how we are using it in _our_ [`smart-home-security-system`](https://github.com/dwyl/smart-home-security-system/blob/85a3357171a8755a7b5e697aca5910b5cbb86cde/getting_started.md#required-software) [project](https://github.com/dwyl/smart-home-auth-server#development) 🏡  

![image](https://user-images.githubusercontent.com/194400/87664760-b6bdcf00-c75d-11ea-8b07-3b12eaba5272.png)

As with any [new technology](https://github.com/dwyl/technology-stack/issues/18) we adopt @dwyl, we need to have a beginner friendly intro explaining exactly what problem the tech/tool solves for us and _why_ it's a total no-brainer to use it instead of anything `else`. We are very conscious/deliberate with our [technology stack](https://github.com/dwyl/technology-stack#why) selection/adoption because we know each addition adds [**cognitive load**](https://en.wikipedia.org/wiki/Cognitive_load), complexity and maintenance overhead. That being said we are _stoked_ when people introduce us to new tools that can help make our lives easier. So ... 

Please use this repository to write up _everything_ you know about [**`insomnia core`**](https://insomnia.rest/products/core) so that someone who has never heard of it before can get started. 
In the "How?" section, it's reasonably safe to assume they know what a [**`REST API`**](https://en.wikipedia.org/wiki/Representational_state_transfer) is (_even if they have never built one themselves_)
and that the have some familiarity with **`Elixir/Phoenix`**. But the tutorial is meant to show them everything they need to know from first principals so don't make any assumptions beyond ["Todo List"](https://github.com/dwyl/phoenix-todo-list-tutorial) level. 

# Why?

Describe _why_ someone would want to use [**`insomnia core`**](https://insomnia.rest/products/core/) for designing/documenting/managing a **`REST API`**. 
Summarise the **key benefits** preferably in bullet points so people can quickly see that it's relevant to them.

> We always [start with _why_](https://github.com/dwyl/start-here#why) because it's the _first_ question people ask themselves. 
Within a handful of seconds of starting to read a page, people think: "_What's in it for **me**_?" (WIIFM)
Not because they are self-centred, just because they are busy and need to know _fast_ if it's worth their time. 
i.e. "_Why should I bother with this when there are so many other things competing for my attention?_"
So we need to _answer_ that question ASAP in anything we write so people don't _immediately_ [**bounce**](https://yoast.com/understanding-bounce-rate-google-analytics/).

# What?

In layperson's terms, what are the key features/benefits of **`insomnia`** over handrolling the API or using a different API creation/management tool? Again, use bullet points or a short paragraph.

## Limitations?

At present **`insomnia` does not support `WebSockets`** https://github.com/Kong/insomnia/issues/528 and there does not appear to be a timeline for it. 
As noted in ["Why `Elixir`"](https://github.com/dwyl/learn-elixir/issues/102) one of the biggest reasons we _use_ Elixir/Phoenix is for Channels. 
Can we still use **`insomnia`** for the **`REST API`** portion and then hand-craft the spec/docs for our Channels/RealTime API?

# Who?

Answer the question: "_**Who** is this tutorial for?_" 
(_it's definitely not "**everyone**" ... but what subset of the dev community is it relevant to?_)

# _How_?

+ [ ] Show how to get started with **insomina** from scratch in a **`new`** project. (_this repo_)
+ [ ] Create a basic API step-by-step

> If you're stuck for a _basic_ example of a REST API to build with insomnia, just build out the endpoints of a Todo List which should already be familiar to the reader so it's less explaining. 
> Or, if you just want an _ultra_ basic API with just an endpoint for `GET` and `POST`,
Consider the "Quotes" example we used in the https://github.com/dwyl/phoenix-content-negotiation-tutorial
Where the `POST` would be adding a `new` quote to the database. 

+ [ ] Demonstrate the reader how **insomnia** makes creating/extending/maintaining the API easier.

<hr />

# tl;dr

Technical writing (_written communication_) is every bit as important as crafting code, read: https://github.com/nelsonic/nelsonic.github.io/issues/797
So please take your time to write something that is a beginner-friendly as possible. 
We have _several_ examples you can follow e.g:
+ https://github.com/dwyl/phoenix-chat-example
+ https://github.com/dwyl/phoenix-todo-list-tutorial
+ https://github.com/dwyl/phoenix-liveview-counter-tutorial

> Once you're done, you could mirror the tutorial to your [_personal_ site](https://github.com/th0mas/tomhaines.uk) as a "blog" post/tutorial. 
When you applied to work with us this [summer](https://github.com/dwyl/hq/issues/556), we looked at your personal site as evidence of your proactiveness. 
And _specifically_ at your [**`Fancy Lights`**](https://github.com/th0mas/Fancy-Lights) "lockdown" project as a sign you could actually get work done. 🚀 
The more content you can _publish_ on your personal site the better for your long-term job/career goals.

**Comment by t0mhaines** [comment]
I'll get on this tomorrow (Friday)

**Comment by nelsonic** [comment]
Thanks @th0mas you're a ⭐ ! 
(love what you've proactively done with [**`learn-scenic`**](https://github.com/dwyl/learn-scenic) so far; keep up the **great work!**) 🚀

**Comment by iteles** [comment]
@nelsonic On a slightly separate note, this description of the 'Why? What? Who? How?' format explanation would be really useful as a readme in our [`contributing`](https://github.com/dwyl/contributing) repo! I can open an issue there.

**Comment by nelsonic** [comment]
@iteles yeah, go for it. re-purpose however you see fit. 👍

**Comment by nelsonic** [comment]
![image](https://user-images.githubusercontent.com/194400/88075245-6df88280-cb70-11ea-90e8-aead57c89ef3.png)

When you first open Insomnia core you will see an empty workspace:

![image](https://user-images.githubusercontent.com/194400/88075291-7d77cb80-cb70-11ea-84a0-ef558e6cc11f.png)

Before:
![Step 1](https://i.imgur.com/O6svSVZ.png)

After:
![step-1-insomnia-create-workspace](https://user-images.githubusercontent.com/194400/88075760-258d9480-cb71-11ea-94cd-d2dbe0ecc53f.png)

Reduce the size of the screenshot to the bare minimum and highlight the important part.

**Comment by nelsonic** [comment]
![insomnia-create-new-request](https://user-images.githubusercontent.com/194400/88076587-296de680-cb72-11ea-9e55-e56fe545c888.png)

![insomnia-new-request-name](https://user-images.githubusercontent.com/194400/88076734-54f0d100-cb72-11ea-9c25-f3a2a6eb42b3.png)

![insomnia-url-send](https://user-images.githubusercontent.com/194400/88076886-87023300-cb72-11ea-8c62-63eb97264914.png)

![insomnia-response](https://user-images.githubusercontent.com/194400/88077059-bd3fb280-cb72-11ea-90e9-64dc9f05dfbc.png)

**Comment by nelsonic** [comment]
The old image had **`8080`** in it from the `Go` example, which was also updated to use TCP Port `4000`:

![step 2.1](https://i.imgur.com/5pmT98Q.png)

**Comment by nelsonic** [comment]
Text refers to left, middle and right columns, so rather than updating the text, just got a new `4000` screenshot:
![insomnia-response-3-columns](https://user-images.githubusercontent.com/194400/88077363-1b6c9580-cb73-11ea-86eb-ab01c8fb959e.png)

**Comment by nelsonic** [comment]
<img width="1020" alt="insomnia-create-todo" src="https://user-images.githubusercontent.com/194400/88079300-6b4c5c00-cb75-11ea-9b9e-1ff96f748bcd.png">

**Comment by nelsonic** [comment]
![insomnia-manage-environments](https://user-images.githubusercontent.com/194400/88096864-3ea53e00-cb8f-11ea-9773-6062532bde3e.png)

![insomnia-create-environment](https://user-images.githubusercontent.com/194400/88096984-6bf1ec00-cb8f-11ea-9061-73366a9e093a.png)

![insomnia-double-click-to-edit-environment-name](https://user-images.githubusercontent.com/194400/88097605-4fa27f00-cb90-11ea-9843-fd6d14e42b87.png)


![insomnia-environment-developmet](https://user-images.githubusercontent.com/194400/88097702-78c30f80-cb90-11ea-8f56-8a21b6152d4e.png)


Before:
![Host Template](https://i.imgur.com/d7BPxi3.png)

After:
![insomnia-host-environment-variable](https://user-images.githubusercontent.com/194400/88097869-c475b900-cb90-11ea-9efb-0d30b74f24c6.png)


Use the development environment:
![insomnia-use-development](https://user-images.githubusercontent.com/194400/88098016-00108300-cb91-11ea-93e0-5a047a88896d.png)

**Comment by nelsonic** [comment]
![insomnia-localhost](https://user-images.githubusercontent.com/194400/88099714-77471680-cb93-11ea-992b-36c7072155fc.png)

![insomnia-host](https://user-images.githubusercontent.com/194400/88099716-78784380-cb93-11ea-85aa-e15c6d4a9c58.png)

**Comment by nelsonic** [comment]
![insomnia-post-request-review-pr](https://user-images.githubusercontent.com/194400/88099944-d016af00-cb93-11ea-888e-90d18ba15dd4.png)

**Comment by nelsonic** [comment]
![insomnia-new-request-view-item](https://user-images.githubusercontent.com/194400/88100580-9f834500-cb94-11ea-8ecb-ea3ec3295de5.png)


<img width="1112" alt="insomnia-response-autocomplete" src="https://user-images.githubusercontent.com/194400/88102724-f0e10380-cb97-11ea-9076-8e4c87385b4d.png">

**Comment by nelsonic** [comment]
<img width="405" alt="insomnia-response-error" src="https://user-images.githubusercontent.com/194400/88102943-41586100-cb98-11ea-88c2-f0934ac38768.png">

**Comment by nelsonic** [comment]
<img width="1214" alt="insomnia-auth-request-authenticated" src="https://user-images.githubusercontent.com/194400/88109683-d9f3de80-cba2-11ea-8115-0a06b04dccf1.png">

**Comment by nelsonic** [comment]
<img width="1112" alt="insomnia-edit-tag" src="https://user-images.githubusercontent.com/194400/88108256-71a3fd80-cba0-11ea-9a94-9fb5b04410d9.png">


<img width="1214" alt="insomnia-send-view-item" src="https://user-images.githubusercontent.com/194400/88108509-e5dea100-cba0-11ea-885b-93cfdd8fcfd1.png">

<img width="1214" alt="insomnia-token-auth" src="https://user-images.githubusercontent.com/194400/88108855-71f0c880-cba1-11ea-83f4-378904411115.png">

<img width="1214" alt="insomnia-auth-request" src="https://user-images.githubusercontent.com/194400/88109167-f6dbe200-cba1-11ea-8159-7e60b4d0eb7d.png">

**Comment by nelsonic** [comment]
PR #3 merged. ✅ 
Thanks again @th0mas 🥇
